### PR TITLE
KafkaV2SinkConnector

### PR DESCRIPTION
**Pull request by xinlian12**
Feature request https://github.com/Azure/azure-sdk-for-java/issues/38769

In this PR, we added the kafka CosmosDB Sink Connector V2 version. 

### Config
- `kafka.connect.cosmos.accountEndpoint` -> No default. Cosmos DB Account Endpoint Uri 
- `kafka.connect.cosmos.accountKey` -> No default. Cosmos DB Account Key         
- `kafka.connect.cosmos.useGatewayMode` -> Default false. Flag to indicate whether to use gateway mode. By default it is false.
- `kafka.connect.cosmos.preferredRegionsList` -> Default empty list. Preferred regions list to be used for a multi region Cosmos DB account. This is a comma separated value (e.g., `[East US, West US]` or `East US, West US`) provided preferred regions will be used as hint. You should use a collocated kafka cluster with your Cosmos DB account and pass the kafka cluster region as preferred region. See list of azure regions [here](https://docs.microsoft.com/dotnet/api/microsoft.azure.documents.locationnames?view=azure-dotnet&preserve-view=true)
-  `kafka.connect.cosmos.applicationName` -> Default empty string. Will be added as the userAgent suffix.  
- `kafka.connect.cosmos.sink.database.name` -> No Default. CosmosDb database name.
- `kafka.connect.cosmos.sink.containers.topicMap` -> No Default. A comma delimited list of Kafka topics mapped to Cosmos containers. For example: topic1#con1,topic2#con2.
- `kafka.connect.cosmos.sink.errors.tolerance` -> Default `None`. Error tolerance level after exhausting all retries. `None` for fail on error. `All` for log and continue
- `kafka.connect.cosmos.sink.bulk.enabled`  -> Default `true`. Flag to indicate whether Cosmos DB bulk mode is enabled for Sink connector. 
- `kafka.connect.cosmos.sink.bulk.maxConcurrentCosmosPartitions` -> Default -1. Usually this is only required to be tuned for large containers. Cosmos DB Item Write Max Concurrent Cosmos Partitions. If not specified it will be determined based on the number of the container's physical partitions which would indicate every batch is expected to have data from all Cosmos physical partitions. If specified it indicates from at most how many Cosmos Physical Partitions each batch contains data. So this config can be used to make bulk processing more efficient when input data in each batch has been repartitioned to balance to how many Cosmos partitions each batch needs to write. This is mainly useful for very large containers (with hundreds of physical partitions.    
- `kafka.connect.cosmos.sink.bulk.initialBatchSize`  -> Default 1. Cosmos DB initial bulk micro batch size - a micro batch will be flushed to the backend when the number of documents enqueued exceeds this size - or the target payload size is met. The micro batch size is getting automatically tuned based on the throttling rate. By default the initial micro batch size is 1. Reduce this when you want to avoid that the first few requests consume too many RUs. 
- `kafka.connect.cosmos.sink.write.strategy` -> Default `ItemOverwrite`. Cosmos DB Item write Strategy: `ItemOverwrite` (using upsert), `ItemAppend` (using create, ignore pre-existing items i.e., Conflicts), `ItemDelete` (deletes based on id/pk of data frame), `ItemDeleteIfNotModified` (deletes based on id/pk of data frame if etag hasn't changed since collecting id/pk), `ItemOverwriteIfNotModified` (using create if etag is empty, update/replace with etag pre-condition otherwise, if document was updated the pre-condition failure is ignored) 
- `kafka.connect.cosmos.sink.maxRetryCount`  -> Default 10. Cosmos DB max retry attempts on write failures for Sink connector. By default, the connector will retry on transient write errors for up to 10 times.
- `kafka.connect.cosmos.sink.id.strategy` -> Default `ProvidedInValueStrategy`. A strategy used to populate the document with an ``id``. Valid strategies are: ``TemplateStrategy``, ``FullKeyStrategy``, ``KafkaMetadataStrategy``, ``ProvidedInKeyStrategy``, ``ProvidedInValueStrategy``. Configuration properties prefixed with``id.strategy`` are passed through to the strategy. For example, when using ``id.strategy=TemplateStrategy`` , the property ``id.strategy.template`` is passed through to the template strategy and used to specify the template string to be used in constructing the ``id``.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-kafka-connect/pom.xml [hunk 1]
```diff
@@ -93,6 +104,11 @@ Licensed under the MIT License.
       <scope>test</scope>
       <version>1.10.0</version> <!-- {x-version-update;org.apache.commons:commons-text;external_dependency} -->
     </dependency>
+    <dependency>
+      <groupId>com.jayway.jsonpath</groupId>
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-kafka-connect/pom.xml:108]: I don't think we would want to add any transitive dependencies that aren't whitelisted by the central team here - so, shading or getting rid of this dependency

File: eng/versioning/external_dependencies.txt [hunk 2]
```diff
@@ -391,6 +391,7 @@ cosmos_org.scalastyle:scalastyle-maven-plugin;1.0.0
 ## Cosmos Kafka connector under sdk\cosmos\azure-cosmos-kafka-connect\pom.xml
 # Cosmos Kafka connector runtime dependencies
 cosmos_org.apache.kafka:connect-api;3.6.0
+cosmos_com.jayway.jsonpath:json-path;2.9.0
```
Comment (FabianMeiswinkel) [inline eng/versioning/external_dependencies.txt:394]: please revert

File: sdk/cosmos/azure-cosmos-kafka-connect/doc/configuration-reference.md [hunk 3]
```diff
@@ -23,3 +23,16 @@
...
...an be used to make bulk processing more efficient when input data in Spark has been repartitioned to balance to how many Cosmos partitions each Spark partition needs to write. This is mainly useful for very large containers (with hundreds of physical partitions.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-kafka-connect/doc/configuration-reference.md:34]: remove spark?

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/KafkaCosmosExceptionsHelper.java [hunk 4]
```diff
@@ -8,7 +8,7 @@
 import org.apache.kafka.connect.errors.ConnectException;
 import org.apache.kafka.connect.errors.RetriableException;
 
-public class CosmosExceptionsHelper {
+public class KafkaCosmosExceptionsHelper {
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/KafkaCosmosExceptionsHelper.java:11]: Why is this renaming necessary - isn't kafka package sufficient?

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/CosmosSinkConfig.java [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        "Flag to indicate whether Cosmos DB bulk mode is enabled for Sink connector. By default it is true.";
+    private static final String BULK_ENABLED_DISPLAY = "enable bulk mode.";
+    private static final boolean DEFAULT_BULK_ENABLED = true;
+
+    // TODO[Public Preview]: Add other write config, for example patch, bulkUpdate
+    public static final String BULK_MAX_CONCURRENT_PARTITIONS_CONF = SINK_CONFIG_PREFIX + "bulk.maxConcurrentCosmosPartitions";
+    private static final String BULK_MAX_CONCURRENT_PARTITIONS_DOC =
+        "Cosmos DB Item Write Max Concurrent Cosmos Partitions."
+            + " If not specified it will be determined based on the number of the container's physical partitions -"
+            + " which would indicate every Spark partition is expected to have data from all Cosmos physical partitions."
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/CosmosSinkConfig.java:40]: Remove Spark

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/KafkaCosmosWriterBase.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            .flatMap(collectionCache -> {
+                return collectionCache
+                    .resolveByNameAsync(
+                        null,
+                        ImplementationBridgeHelpers
+                            .CosmosAsyncContainerHelper
+                            .getCosmosAsyncContainerAccessor()
+                            .getLinkWithoutTrailingSlash(container),
+                        null,
+                        new DocumentCollection())
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/KafkaCosmosWriterBase.java:111]: why new DocumentCollection and not null? I think his param is meant to be used when asynccache has an instance believed to be stale?

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/idstrategy/TemplateStrategyConfig.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        this(getConfig(), props);
+    }
+
+    public TemplateStrategyConfig(ConfigDef definition, Map<String, ?> originals) {
+        super(definition, originals);
+
+        this.template = getString(TEMPLATE_CONFIG);
+    }
+
+
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/idstrategy/TemplateStrategyConfig.java:32]: NIT remove one lf

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/source/CosmosChangeFeedModes.java [hunk 8]
```diff
@@ -18,7 +18,7 @@ public String getName() {
 
     public static CosmosChangeFeedModes fromName(String name) {
         for (CosmosChangeFeedModes mode : CosmosChangeFeedModes.values()) {
-            if (mode.getName().equals(name)) {
+            if (mode.getName().equalsIgnoreCase(name)) {
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/source/CosmosChangeFeedModes.java:21]: why - we always write it - why would we need to be case-insensitive?

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM

File: sdk/cosmos/azure-cosmos-kafka-connect/pom.xml (hunk 1 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/pom.xml:108]: Added shading

File: eng/versioning/external_dependencies.txt (hunk 2 above)
Comment (xinlian12) [reply eng/versioning/external_dependencies.txt:395]: Added shading

File: sdk/cosmos/azure-cosmos-kafka-connect/doc/configuration-reference.md (hunk 3 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/doc/configuration-reference.md]: updated, changed from `spark` -> `batch`

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/KafkaCosmosExceptionsHelper.java (hunk 4 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/KafkaCosmosExceptionsHelper.java:11]: mostly just easier to differentiate between same class file between different projects. 
for example, `CosmosConstants -> KafkaCosmosContants. `

But no strong opinion, can revert back

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/KafkaCosmosWriterBase.java (hunk 6 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/KafkaCosmosWriterBase.java:111]: both works. The value used here just mean if the cached value is not the same as the staled value, then use the cached value directly. So null or new DocumentCollection() would have the same effect

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/idstrategy/TemplateStrategyConfig.java (hunk 7 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/idstrategy/TemplateStrategyConfig.java]: removed

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/source/CosmosChangeFeedModes.java (hunk 8 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/source/CosmosChangeFeedModes.java:21]: either "allVersionxxx" or "ALLVERSIONxxx" should all be considered valid, so changed to use ignoreCase here

File: sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/KafkaCosmosWriterBase.java (hunk 6 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/KafkaCosmosWriterBase.java:111]: BUt null is cheaper (no new garbage instantiation)

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/sink/KafkaCosmosWriterBase.java:111]: will change in next PR

**Pull request merged**